File: fixtures/scale/csharp/medium/Functions/FunctionRegistry.cs
using System;
using System.Collections.Generic;
using Calc.Eval;

namespace Calc.Functions
{
    public class FunctionRegistry
    {
        private readonly Dictionary<string, Func<double[], double>> _table =
            new Dictionary<string, Func<double[], double>>();

        public void Define(string name, Func<double[], double> body)
        {
            _table[name] = body;
        }

        public double Invoke(string name, double[] arguments)
        {
            if (!_table.TryGetValue(name, out var body))
            {
                throw new EvalError("unknown function", name);
            }
            return body(arguments);
        }

        public List<string> Names()
        {
            var names = new List<string>(_table.Keys);
            names.Sort();
            return names;
        }
    }
}

File: fixtures/scale/php/medium/src/Functions/FunctionRegistry.php
<?php

namespace Calc\Functions;

use Calc\Eval\EvalError;

class FunctionRegistry
{
    /** @var array<string, callable> */
    private array $table = [];

    public function define(string $name, callable $body): void
    {
        $this->table[$name] = $body;
    }

    public function invoke(string $name, array $arguments): float
    {
        if (!isset($this->table[$name])) {
            throw new EvalError('unknown function', $name);
        }
        return ($this->table[$name])(...$arguments);
    }

    public function names(): array
    {
        $names = array_keys($this->table);
        sort($names);
        return $names;
    }
}
